File: src/main/java/com/snackgame/server/applegame/domain/Board.java
package com.snackgame.server.applegame.domain;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import com.snackgame.server.applegame.domain.exception.AppleNotRemovableException;
import com.snackgame.server.applegame.domain.exception.InvalidRangeException;

public class Board {

    private static final int REMOVABLE_SUM = 10;

    private final List<List<Apple>> apples;

    public Board(List<List<Apple>> apples) {
        this.apples = apples.stream()
                .map(ArrayList::new)
                .collect(Collectors.toList());
    }

    public static Board ofRandomized(int height, int width) {
        List<List<Apple>> apples = new ArrayList<>();
        for (int i = 0; i < height; i++) {
            List<Apple> row = new ArrayList<>();
            for (int j = 0; j < width; j++) {
                row.add(Apple.ofRandomizedNumber());
            }
            apples.add(row);
        }
        return new Board(apples);
    }

    public int removeApplesIn(Range range) {
        validateContains(range);
        validateSumOf(range);
        int removed = 0;
        for (Coordinate coordinate : range.getAppleCoordinates()) {
            removeAppleAt(coordinate);
            ++removed;
        }
        return removed;
    }

    private void validateContains(Range range) {
        boolean appleDoesNotExist = getApplesIn(range.getAppleCoordinates()).stream()
                .anyMatch(Apple::isEmpty);
        boolean unexpectedAppleExists = getApplesIn(range.getEmptyCoordinates()).stream()
                .anyMatch(Apple::exists);
        if (appleDoesNotExist || unexpectedAppleExists) {
            throw new InvalidRangeException();
        }
    }

    private void validateSumOf(Range range) {
        if (sumApplesIn(range) != REMOVABLE_SUM) {
            throw new AppleNotRemovableException("사과들의 합이 " + REMOVABLE_SUM + "이 아닙니다");
        }
    }

    private int sumApplesIn(Range range) {
        return getApplesIn(range.getAppleCoordinates()).stream()
                .map(Apple::getNumber)
                .reduce(0, Integer::sum);
    }

    private List<Apple> getApplesIn(List<Coordinate> coordinates) {
        return coordinates.stream()
                .map(coordinate -> apples.get(coordinate.getY()).get(coordinate.getX()))
                .collect(Collectors.toList());
    }

    private void removeAppleAt(Coordinate coordinate) {
        List<Apple> row = apples.get(coordinate.getY());
        row.set(coordinate.getX(), Apple.EMPTY);
    }

    public List<List<Apple>> getApples() {
        return apples;
    }

    @Override
    public String toString() {
        return "Board{" +
                "numbers=" + apples +
                '}';
    }
}

File: src/main/java/com/snackgame/server/game/snackgame/core/domain/Board.kt
package com.snackgame.server.game.snackgame.core.domain

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.snackgame.server.game.snackgame.core.domain.snack.EmptySnack
import com.snackgame.server.game.snackgame.core.domain.snack.Snack
import com.snackgame.server.game.snackgame.exception.InvalidBoardSizeException
import com.snackgame.server.game.snackgame.exception.InvalidCoordinateException
import com.snackgame.server.game.snackgame.exception.SnackNotRemovableException
import java.util.stream.Collectors

@JsonIgnoreProperties("height", "width")
class Board() {
    private var snacks: MutableList<MutableList<Snack>> = arrayListOf()

    constructor(snacks: MutableList<MutableList<Snack>>) : this() {
        validateIsRectangle(snacks)
        this.snacks = snacks.map { ArrayList(it) }.toMutableList()
    }

    constructor(height: Int, width: Int) : this(createRandomized(height, width))

    private fun validateIsRectangle(snacks: List<List<Snack>>) {
        if (snacks.isEmpty() || snacks[0].isEmpty()) {
            throw InvalidBoardSizeException()
        }
    }

    fun reset(): Board {
        return Board(createRandomized(height, width))
    }

    fun removeSnacksIn(streak: Streak): List<Snack> {
        validateIsIncluded(streak.coordinates)
        validateSumOf(streak.coordinates)

        val snacksToRemove = streak.coordinates.map { snacks[it.y][it.x] }
        streak.coordinates.forEach { removeSnacksAt(it) } // TODO: 제거한 스낵과 제거할 스낵 사이에 스낵이 없음을 검증해야 한다
        return snacksToRemove
    }

    fun bombSnacksIn(streak: Streak): List<Snack> {
        val filteredStreak = filterValidCoordinates(streak.coordinates)

        val snacksToRemove = filteredStreak.map { snacks[it.y][it.x] }
            .filter { it !is EmptySnack }
        filteredStreak.forEach { removeSnacksAt(it) } // TODO: 제거한 스낵과 제거할 스낵 사이에 스낵이 없음을 검증해야 한다
        return snacksToRemove
    }

    private fun validateIsIncluded(coordinates: List<Coordinate>) {
        coordinates.forEach { coordinate ->
            if (coordinate.y >= snacks.size || coordinate.x >= snacks[0].size) {
                throw InvalidCoordinateException()
            }
        }
    }

    private fun validateSumOf(coordinates: List<Coordinate>) {
        if (sumSnacksIn(coordinates) != REMOVABLE_SUM) {
            throw SnackNotRemovableException("스낵들의 합이 " + REMOVABLE_SUM + "이 아닙니다")
        }
    }

    private fun sumSnacksIn(coordinates: List<Coordinate>): Int {
        return getSnacksIn(coordinates).stream().map(Snack::getNumber).reduce(0, Integer::sum)

    }

    private fun getSnacksIn(coordinates: List<Coordinate>): List<Snack> {
        return coordinates.stream().map { coordinate -> snacks[coordinate.y][coordinate.x] }
            .collect(Collectors.toList())
    }

    private fun removeSnacksAt(coordinate: Coordinate): Snack {
        val removed = snacks[coordinate.y][coordinate.x]
        snacks[coordinate.y][coordinate.x] = EmptySnack.get()
        return removed
    }

    private fun filterValidCoordinates(coordinates: List<Coordinate>): List<Coordinate> {
        return coordinates.filter { isInside(it) }
    }

    private fun isInside(coordinate: Coordinate): Boolean {
        return coordinate.y in snacks.indices && coordinate.x in snacks[0].indices
    }

    fun getSnacks(): List<List<Snack>> {
        return snacks.map { ArrayList(it) }
    }

    private val height get() = snacks.size
    private val width get() = snacks[0].size

    companion object {
        private const val REMOVABLE_SUM = 10
    }
}
